python/flink_agents/api/memory/long_term_memory.py

```diff
@@ -0,0 +1,294 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+#################################################################################
+import importlib
+from abc import ABC, abstractmethod
+from datetime import datetime
+from enum import Enum
+from typing import Any, Dict, List, Type
+
+from pydantic import (
+    BaseModel,
+    Field,
+    field_serializer,
+    model_validator,
+)
+from typing_extensions import override
+
+from flink_agents.api.chat_message import ChatMessage
+from flink_agents.api.prompts.prompt import Prompt
+
+ItemType = str | ChatMessage
+
+
+class CompactionStrategyType(Enum):
+    """Strategy for compact memory set."""
+
+    SUMMARIZATION = "summarization"
+
+
+class CompactionStrategy(BaseModel, ABC):
+    """Strategy for compact memory set."""
+
+    @property
+    @abstractmethod
+    def type(self) -> CompactionStrategyType:
+        """Return type of this strategy."""
+
+
+class SummarizationStrategy(CompactionStrategy):
+    """Summarization strategy."""
+
+    model: str
+    prompt: str | Prompt | None = None
+
+    @property
+    @override
+    def type(self) -> CompactionStrategyType:
+        return CompactionStrategyType.SUMMARIZATION
+
+
+class LongTermMemoryBackend(Enum):
+    """Backend for Long-Term Memory."""
+
+    VectorStore = "vectorstore"
+
+
+class DatetimeRange(BaseModel):
+    """Represents a datetime range."""
+
+    start: datetime
+    end: datetime
+
+
+class MemorySetItem(BaseModel):
+    """Represents a long term memory item retrieved from vector store.
+
+    Attributes:
+        memory_set_name: The name of the memory set this item belongs to.
+        id: The id of this item.
+        value: The value of this item.
+        compacted: Whether this item has been compacted.
+        created_time: The timestamp this item was added to the memory set.
+        last_accessed_time: The timestamp this item was last accessed.
+        additional_metadata: Additional metadata for this item.
+    """
+
+    memory_set_name: str
+    id: str
+    value: Any
+    compacted: bool = False
+    created_time: datetime | DatetimeRange = None
+    last_accessed_time: datetime
+    additional_metadata: Dict[str, Any] | None = None
+
+
+class MemorySet(BaseModel):
+    """Represents a long term memory set contains memory items.
+
+    Attributes:
+        name: The name of this memory set.
+        item_type: The type of items stored in this set.
+        capacity: The capacity of this memory set.
+        compaction_strategy: Compaction strategy and additional arguments used
+        to compact memory set.
+        size: The size of this memory set.
+    """
+
+    name: str
+    item_type: Type[str] | Type[ChatMessage]
+    capacity: int
+    compaction_strategy: CompactionStrategy
+    size: int = Field(default=0, exclude=True)
```

Comment:
It's fragile to maintain the size separately from the underlying actual collection. We might consider get the size from the store.